In [7]:
# вбудовані
a = input("Введіть число: ")
b = input("Введіть число: ")
print(f"Ви ввели: {a+b}")

# виведе конкатановано два рядки (просто склеїть)

Ви ввели: 12


In [8]:
# приведення типів

a = int(input("Введіть число: "))
b = int(input("Введіть число: "))
print(f"Ви ввели: {a+b}")

# виведе як суму

Ви ввели: 3


In [13]:
a, b = input("Введіть два числа через пробіл: ").split()
print(f"Ви ввели: {int(a)+int(b)}")

Ви ввели: 3


In [14]:
if a.isdigit() and b.isdigit():
    print(f"Ви ввели: {int(a)+int(b)}")

Ви ввели: 3


In [15]:
a=2
if a==2:
    print(a)
else:
    print(f"not {a}")

2


In [18]:
a = None
if a:
    print(a)
else:
    print(f"not defined")

not defined


In [20]:
a = input("Введіть a: ")
if a: # None, 0, "", [], {}, () - розпізнається як False
    print(a)
else:
    print(f"not defined")

not defined


In [2]:
a=2
a != 2

False

In [3]:
a>1

True

In [4]:
a<1

False

In [5]:
3>a>1

True

In [7]:
3>a>2 #якщо хоч одне не виконується, то вираз не виконується

False

In [8]:
# це аналог конструкції and
3>a and a>2

False

In [9]:
a>=2

True

In [11]:
# це аналог конструкції (застаріло)
not(a<2)

True

In [12]:
-3 **3 #пріоритети виконання операцій - як в математиці унарний мінус нижче приорітет ніж піднесення до ступеню

-27

In [14]:
a = 0.2 + 0.1
print(a)

# виводить 0.30000000000000004 бо використовується двійкова система числення і вона не може точно 
# представити десяткові дроби (особливість процесора і пам'яті)
# Ця ситуація може статися незалежно від мови програмування, і пов'язана вона з внутрішнім представленням чисел комп'ютером. 
# Тому порівняння дійсних чисел можливе лише з певною точністю, наприклад, до шостого знака після коми. 
# Це коли порівнюють між собою не два числа, а їх різницю з точністю до 0.000001, та перевіряють: вона менше точності чи ні. 
# Якщо при обчисленнях точність стає важливою, програмісти на Python використовують спеціальний модуль Decimal. 

0.30000000000000004


In [16]:
age = 18
is_adult = age >= 18  # True
is_adult

True

In [17]:
age = 15
is_adult = age >= 18  # False
is_adult

False

Розглянемо знайомий всім приклад зі школи. 
Нехай у нас є прямокутний трикутник зі сторонами 10 і 5. Позначимо сторони наступними змінними:

In [18]:
side_a = 10
side_b = 5
hypotenuse = (side_a**2 + side_b**2) ** 0.5
hypotenuse

11.180339887498949

In [19]:
S = side_a * side_b / 2
S

25.0

Після розрахунку змінна S буде дорівнювати 25.0. Зверни увагу, що змінна S стала типу float, хоча в правій стороні виразу стоять тільки числа int. Так завжди буває при операціях ділення.



Щоб отримати тільки цілу частину, треба використовувати операцію ділення без остачі //. Коли така операція може знадобиться? Розглянемо приклад, де треба скласти програму, яка переводить час із секунд, визначаючи повну кількість годин, хвилин і секунд.



In [21]:
n = 5000

hours = n // (60 * 60)
minutes = (n - hours * 60 * 60) // 60
seconds = n - hours * 60 * 60 - minutes * 60
print(f"{hours}:{minutes}:{seconds}")

1:23:20


Перерахуємо основні правила застосування операцій у порядку їх пріоритету:



- Дужки (): Вирази всередині дужок мають найвищий пріоритет і виконуються першими. Дужки можуть бути використані для зміни звичайного порядку виконання.
- Піднесення до степеня **: Наступним виконується піднесення до степеня.
- Унарні плюс + та мінус -. Далі йдуть операції унарного плюса та мінуса (наприклад, -5 або +3).
- Множення ``, ділення /, цілочисельне ділення //, остача від ділення %: Ці операції мають однаковий пріоритет і виконуються після унарних операцій.
- Додавання + та віднімання -. Виконуються після операцій множення та ділення.
- Операції порівняння ==, !=, >, <, >=, <=: Виконуються після арифметичних операцій.
- Логічні операції not, and, or: В кінці виконуються логічні операції, де not має вищий пріоритет, ніж and, який, у свою чергу, має вищий пріоритет, ніж or.

Розглянемо приклад. Складемо програму, яка обчислює відстані між точками (відстані Евкліда)

In [23]:
x1 = 10
y1 = 10
x2 = 25
y2 = 25
d = ((x2 - x1) ** 2 + (y2 - y1) ** 2) ** 0.5
d

21.213203435596427


Спочатку виконуються операції в дужках. У нашому випадку це x2 - x1 та y2 - y1. Вони відповідають за обчислення різниці між відповідними координатами двох точок.

Далі виконується піднесення до квадрата **2 та застосовується до результатів (x2 - x1) і (y2 - y1), обчислених на попередньому кроці.

Після обчислення квадратів різниць виконується додавання: сума (x2 - x1)**2 і (y2 - y1)**2, бо вона знаходиться в дужках.

Останньою операцією у внутрішніх виразах є піднесення суми квадратів до степеня 0.5, що еквівалентно взяттю квадратного кореня. Це дає кінцевий результат — відстань між двома точками, яка в нашому випадку дорівнює 21.213203435596427.

## Колекції

Перерахуємо основні типи колекцій у Python:



- Списки (Lists). Впорядковані колекції, які можуть містити елементи будь-якого типу. Списки є змінними, що означає можливість зміни, додавання або видалення елементів після їх створення.
- Кортежі (Tuples). Схожі на списки, але вони незмінні. Це означає, що після створення кортежу його не можна змінювати. Кортежі часто використовуються для зберігання даних, які не повинні змінюватися впродовж роботи програми.
- Словники (Dictionaries). Колекції пар ключ-значення, де кожний ключ є унікальним. Словники використовуються, коли потрібен швидкий доступ до елементів за ключем.
- Множини (Sets). Неупорядковані колекції унікальних елементів. Використовуються для видалення дублікатів і виконання операцій, які типові для математичних множин, таких як об'єднання, перетин, різниця.
- Заморожені множини (Frozen Sets). Це незмінні версії множин. Такі множини не можна змінювати після їх створення.

In [25]:
my_list = list()

empty_list = []

my_list = [1, 2, 3, 4, 5]
print(my_list)
my_list = [1, "Hello", 3.14]
print(my_list)
my_list.append(4)
print(my_list)
my_list.remove("Hello")
print(my_list)

[1, 2, 3, 4, 5]
[1, 'Hello', 3.14]
[1, 'Hello', 3.14, 4]
[1, 3.14, 4]


In [27]:
some_iterable = ["a", "b", "c"]
first_letter = some_iterable[0]
middle_one = some_iterable[1]
last_letter = some_iterable[2]
print(first_letter, middle_one, last_letter)

a b c


In [29]:
print(some_iterable[-3])
print(some_iterable[-2])
print(some_iterable[-1])

a
b
c


In [31]:
some_iterable[1] = -2
print(some_iterable)


['a', -2, 'c']


In [32]:
chars = ["a", "b", "c"]
last = chars.pop(1)

In [34]:
chars = ["a", "b"]
print(chars)
numbers = [1, 2]
chars.extend(numbers)
print(chars)
chars.insert(1, "b")
print(chars)

['a', 'b']
['a', 'b', 1, 2]
['a', 'b', 'b', 1, 2]


In [36]:
chars = ["a", "b", "c", "d"]
c_ind = chars.index("c")
print(c_ind)

2


In [37]:
my_list = [1, 2, 3, 4, 2, 2, 5, 2]
count_2 = my_list.count(2)
print(count_2)  # Виведе 4, оскільки число 2 зустрічається 4 рази

4


In [38]:
my_list = [1, 2, 3, 4, 5]
print(len(my_list))

5


In [39]:
nums = [3, 1, 4, 1, 5, 9, 2]
nums.sort()
print(nums)  # Виведе [1, 1, 2, 3, 4, 5, 9]

[1, 1, 2, 3, 4, 5, 9]


In [40]:
nums = [3, 1, 4, 1, 5, 9, 2]
nums.sort(reverse=True)
print(nums)  # Виведе [9, 5, 4, 3, 2, 1, 1]

[9, 5, 4, 3, 2, 1, 1]


In [41]:
words = ["banana", "apple", "cherry"]
words.sort(key=len)
print(words)  # Виведе ['apple', 'banana', 'cherry']

['apple', 'banana', 'cherry']


Вбудований метод sorted() у Python використовується для сортування колекцій. Він відрізняється від методу sort(), який застосовується безпосередньо до списку, та змінює його: sorted() повертає новий відсортований об'єкт, залишаючи вихідний об'єкт без змін. Метод також використовує аргумент reverse=True, щоб відсортувати об'єкти у зворотному порядку, та за допомогою аргументу key можна вказати функцію, яка буде застосовуватися для визначення порядку сортування.


In [42]:
nums = [3, 1, 4, 1, 5, 9, 2]
sorted_nums = sorted(nums)
print(sorted_nums)  # Виведе [1, 1, 2, 3, 4, 5, 9]

[1, 1, 2, 3, 4, 5, 9]


In [43]:
sorted_nums_desc = sorted(nums, reverse=True)
print(sorted_nums_desc)  # Виведе [9, 5, 4, 3, 2, 1, 1]

[9, 5, 4, 3, 2, 1, 1]


In [44]:
words = ["banana", "apple", "cherry"]
sorted_words = sorted(words, key=len)
print(sorted_words)  # Виведе ['apple', 'banana', 'cherry']

['apple', 'banana', 'cherry']


In [45]:
chars = ["a", "b"]
chars_copy = chars.copy()

In [46]:
my_dict = {"name": "Alice", "age": 25, "city": "New York"}
print(my_dict["name"])  # Виведе 'Alice'

Alice


In [47]:
my_dict["age"] = 26  # Змінює вік на 26
my_dict["email"] = "alice@example.com"  # Додає нову пару ключ-значення
print(my_dict)

{'name': 'Alice', 'age': 26, 'city': 'New York', 'email': 'alice@example.com'}


In [48]:
del my_dict["age"]
print(my_dict)

{'name': 'Alice', 'city': 'New York', 'email': 'alice@example.com'}


In [49]:
print("name" in my_dict)
print("age" in my_dict)

True
False


In [51]:
my_dict = {"name": "Alice", "age": 25}
age = my_dict.get("age")  # Поверне 25
gender = my_dict.get("gender")  # Поверне None, оскільки "gender" немає в словнику
print(age)

25


In [52]:
a = set("hello")

In [53]:
#Приклад. Нехай у нас є список, з якого треба видалити дублікати

lst = [1, 2, 3, 1, 2, 2, 3, 4, 1]
lst = list(set(lst))
print(lst)  # Виведе [1, 2, 3, 4]



[1, 2, 3, 4]


In [54]:
numbers = {1, 2, 3}
numbers.add(4)
print(numbers)  # {1, 2, 3, 4}

{1, 2, 3, 4}


In [55]:
numbers = {1, 2, 3}
numbers.remove(3)
print(numbers)  # {1, 2}

{1, 2}


In [56]:
numbers = {1, 2, 3}
numbers.discard(2)
print(numbers)  # {1, 3}

{1, 3}


In [58]:
a = {1, 2, 3}

b = {3, 4, 5}

In [60]:
# Перетин двох множин включає лише ті елементи, які є в обох множинах. Щоб знайти загальні елементи для двох множин, над ними треба виконати операцію & або використати метод intersection:


print(a.intersection(b))  # {3}
print(a & b)  # {3}

{3}
{3}


In [62]:
# Різниця між двома множинами включає елементи, які містяться в першій множині, але не містяться в другій. Можна також використовувати оператор - або метод difference:

a = {1, 2, 3}
b = {3, 4, 5}
print(a.difference(b))  # {1, 2}
print(a - b)  # {1, 2}

{1, 2}
{1, 2}


Симетрична різниця між двома множинами включає всі елементи, які містяться в одній множині, але не містяться в іншій, і навпаки. Щоб знайти всі елементи з двох множин, окрім загальних, нам необхідно використати оператор ^.



Альтернативно, можна використовувати метод symmetric_difference:

In [63]:
a = {1, 2, 3}
b = {3, 4, 5}
print(a.symmetric_difference(b))  # {1, 2, 4, 5}
print(a ^ b)  # {1, 2, 4, 5}

{1, 2, 4, 5}
{1, 2, 4, 5}


Об'єднання двох множин включає всі елементи з обох множин, але без дублікатів. Це знаходиться за допомогою оператора | або методу union:


In [64]:
a = {1, 2, 3}
b = {3, 4, 5}
print(a.union(b))  # {1, 2, 3, 4, 5}
print(a | b)  # {1, 2, 3, 4, 5}

{1, 2, 3, 4, 5}
{1, 2, 3, 4, 5}


Заморожені множини в Python, відомі як frozenset, є подібними до звичайних множин set, але з ключовою відмінністю: вони є незмінними. Це означає, що після створення замороженої множини ви не можете додати або видалити елементи з неї.



Заморожену множину можна створити за допомогою функції frozenset():

In [65]:
my_frozenset = frozenset([1, 2, 3, 4, 5])

In [66]:
a = frozenset([1, 2, 3])
b = frozenset([3, 4, 5])

union = a | b  # Об'єднання множин
intersection = a & b  # Перетин множин
difference = a - b  # Різниця множин
symmetric_difference = a ^ b  # Симетрична різниця

print(union)  # frozenset({1, 2, 3, 4, 5})
print(intersection)  # frozenset({3})
print(difference)  # frozenset({1, 2})
print(symmetric_difference)  # frozenset({1, 2, 4, 5})

frozenset({1, 2, 3, 4, 5})
frozenset({3})
frozenset({1, 2})
frozenset({1, 2, 4, 5})


In [67]:
my_tuple = (1, "Hello", 3.14)

In [69]:
first_item = my_tuple[0]  # Отримати перший елемент
print(first_item)  # 1

1


In [74]:
#Наприклад, розглянемо набір точок на площині (кортежі). Їх можна використовувати як ключі у словнику:


points = { 
    (0, 0): "O",
    (1, 1): "A",
    (2, 2): "B",
}

# Методи рядків


Незмінна послідовність означає, що якщо рядок уже створений, то змінити його не можна, можна тільки створити новий.

s = "Hello world!"
s[0] = "Q"# Тут буде викликано виняток (помилка) TypeError

In [75]:
s = "Hello"
s.upper()
print(s)  # Виведе 'HELLO'

Hello


In [76]:
s = "Some Text"
print(s.lower())  # Виведе 'some text'

some text


In [77]:
s = "Bill Jons"
print(s.startswith("Bi"))  # Виведе True

True


In [78]:
s = "hello.jpg"
print(s.endswith("jpg"))  # Виведе True

True


In [81]:
s = "hello world".capitalize()  # Результат: "Hello world"
s

'Hello world'

In [82]:
s = "hello world".title()  # Результат: "Hello World"
s

'Hello World'

Ще можуть стати в пригоді методи isdigit, isalpha, isspace, які перевіряють, чи складається рядок тільки з цифр, літер, пробілів тощо відповідно:

In [84]:
print("123".isdigit())  # True
print("hello".isalpha())  # True
print(" ".isspace())  # True

True
True
True


In [85]:
# Просте форматування рядка
name = "John"
print("Hello, {}!".format(name))

# Форматування з декількома аргументами
age = 25
print("Hello, {}. You are {} years old.".format(name, age))

# Використання іменованих аргументів
print("Hello, {name}. You are {age} years old.".format(name="Jane", age=30))

# Використання індексів для вказівки порядку аргументів
print("Hello, {1}. You are {0} years old.".format(age, name))

Hello, John!
Hello, John. You are 25 years old.
Hello, Jane. You are 30 years old.
Hello, John. You are 25 years old.


# Зрізи у Python (Slice)

Для впорядкованих контейнерів є спеціальний синтаксис, коли нам необхідно отримати деяку послідовність елементів з контейнера.

Зрізи (slices) у Python — це потужний механізм для доступу до частин послідовностей, таких як рядки, списки та кортежі.

Зрізи визначаються за допомогою квадратних дужок [] із вказівкою індексів початку, кінця та (необов'язково) кроку. Ось основний синтаксис:

послідовність[початок:кінець:крок]

- початок — індекс елемента, з якого починається зріз. Якщо він не вказаний, зріз починається з початку послідовності, з 0.
- кінець — індекс елемента, до якого йде зріз, але увага!, не включаючи його. Якщо він не вказаний, зріз іде до кінця послідовності.
- крок — визначає крок, з яким вибираються елементи. Якщо не вказаний, використовується крок 1.

In [86]:
s = "Hello, World!"
first_five = s[:5]
print(first_five)  # Виведе 'Hello'

Hello


In [88]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
odd_numbers = numbers[0:10:2]
print(odd_numbers)  # Виведе [1, 3, 5, 7, 9]

[1, 3, 5, 7, 9]


In [90]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
even_numbers = numbers[1:10:2]
print(even_numbers)  # Виведе [2, 4, 6, 8, 10]

[2, 4, 6, 8, 10]


In [91]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
reverse_numbers = numbers[::-1]
print(reverse_numbers)

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]


In [92]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
copy_numbers = numbers[:]
print(copy_numbers)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [115]:
num = 15  # приклад значення для num

if num > 10:
    print("num більше за 10")
else:
    print("num не більше за 10")

num більше за 10


In [116]:
x = int(input("Введіть число: "))

if x % 2 == 0:
    print("Число x є парним.")
else:
    print("Число x є непарним.")

Число x є парним.


In [117]:
a = input("Введіть число")
a = int(a)
if a > 0:
    print("Число додатне")
elif a < 0:
    print("Число від'ємне")
else:
    print("Це число - нуль")

Число додатне


In [118]:
a = input("Введіть число")
a = int(a)

if a == 1:
    print("Число дорівнює 1")
elif a > 0:
    print("Число додатне")
else:
    print("a <= 0")

Число дорівнює 1


In [120]:
money = 0
if money:
    print(f"You have {money} on your bank account")
else:
    print("You have no money and no debts")

You have no money and no debts


In [121]:
result = None
if result:
    print(result)
else:
    print("Result is None, do something")

Result is None, do something


In [122]:
user_name = input("Enter your name: ")

if user_name:
    print(f"Hello {user_name}")
else:
    print("Hi Anonym!")

Hi Anonym!


In [123]:
a = [1, 2, 3]
b = a
c = [1, 2, 3]

print(a is b)  # True
print(a is c)  # False

True
False


Що робити, якщо у нас складна умова, яка поєднує в собі декілька вкладених умов? Наприклад, щоб користувач міг орендувати автомобіль, потрібно, щоб у користувача обов'язково було вказане ім'я, користувач був старший 18 і були водійські права.

In [124]:
name = "Taras"
age = 22
has_driver_licence = True

if name and age >= 18 and has_driver_licence:
    print(f"User {name} can rent a car")

User Taras can rent a car


In [125]:
num = int(input("Введіть число: "))

length = len(str(num))

if length == 2 and num % 2 == 0:
    print("Парне двозначне число")
else:
    print("Ні")

Ні


In [126]:
# Задаємо конкретне число
num = int(input())

# Перевіряємо кратність
if num % 3 == 0 and num % 5 == 0:
    print("FizzBuzz")
elif num % 3 == 0:
    print("Fizz")
elif num % 5 == 0:
    print("Buzz")
else:
    print(num)

34


In [127]:
x = int(input("X: "))
y = int(input("Y: "))

if x == 0:
    print("X can`t be equal to zero")
    x = int(input("X: "))

    if x == 0:
        print("X can`t be equal to zero")
        x = int(input("X: "))

        if x == 0:
            print("X can`t be equal to zero")
            x = int(input("X: "))

result = y / x

X can`t be equal to zero


Наведемо наступний приклад вкладеності для визначення чвертей для координатної площини, щоб закріпити розуміння блоку інструкцій. Даний фрагмент коду реалізує визначення квадранту точки з координатами x та y на координатній площині.



In [128]:
if x >= 0:
    if y >= 0:  # x > 0, y > 0
        print("Перша чверть")
    else:  # x > 0, y < 0
        print("Четверта чверть")
else:
    if y >= 0:  # x < 0, y > 0
        print("Друга чверть")
    else:  # x < 0, y < 0
        print("Третя чверть")

Перша чверть


In [130]:
is_nice = True
if is_nice:
    state = "nice"
else:
    state = "not nice"
print(state)    

nice


In [132]:
is_nice = True
state = "nice" if is_nice else "not nice"
print(state)

nice


In [133]:
some_data = None
msg = some_data or "Не було повернено даних"
print(msg)

Не було повернено даних


In [134]:
fruit = "apple"

match fruit:
    case "apple":
        print("This is an apple.")
    case "banana":
        print("This is a banana.")
    case "orange":
        print("This is an orange.")
    case _:
        print("Unknown fruit.")

This is an apple.


In [136]:
point = (0,4)
match point:
    case (0, 0):
        print("Точка в центрі координат")
    case (0, y):
        print(f"Точка лежить на осі Y: y={y}")
    case (x, 0):
        print(f"Точка лежить на осі X: x={x}")
    case (x, y):
        print(f"Точка має координати:  x={x}, y={y}")
    case _:
        print("Це не точка")

Точка лежить на осі Y: y=4


In [138]:
pets = ["dog", "fish", "cat"]

match pets:
    case ["dog", "cat", _]:
        # Випадок, коли є і собака, і кіт
        print("There's a dog and a cat.")
    case ["dog", _, _]:
        # Випадок, коли є тільки собака
        print("There's a dog.")
    case _:
        # Випадок для інших комбінацій
        print("No dogs.")

# В цьому випадку виведення буде "There's a dog.", бо тільки "dog" знаходиться на своєму місці.

There's a dog.


In [139]:
alphabet = "abcdefghijklmnopqrstuvwxyz"
for char in alphabet:
    print(char, end=" ")

a b c d e f g h i j k l m n o p q r s t u v w x y z 

In [140]:
# Зчитування рядка від користувача
user_input = input("Введіть рядок: ")

# Ініціалізація змінних для підрахунку символів та пробілів
total_chars = len(user_input)  # загальна кількість символів у рядку
space_count = 0  # кількість пробілів

# Підрахунок кількості пробілів
for char in user_input:
    if char == " ":
        space_count += 1

# Виведення результатів
print(f"Загальна кількість символів у рядку: {total_chars}")
print(f"Кількість пробілів у рядку: {space_count}")

Загальна кількість символів у рядку: 17
Кількість пробілів у рядку: 3


In [142]:
k = 0
while k < 10:
    k = k + 1
    print(k)

1
2
3
4
5
6
7
8
9
10


In [143]:
a = 0
while True:
    print(a)
    if a >= 20:
        break
    a = a + 1

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20


In [144]:
a = 0
while a < 6:
    a = a + 1
    if not a % 2:
        continue
    print(a)

# Інструкція print(a) не виконувалась, коли a ділилося на 2 без залишку, оскільки ітерація завершувалася за допомогою continue.

1
3
5


# Розширення можливостей циклу for


- Функція range важлива для створення послідовностей чисел, які ви можете використовувати у циклах. Вона надзвичайно корисна, коли вам потрібно виконати дію певну кількість разів або ітерувати через послідовність чисел.
- Коли вам потрібно отримати доступ не тільки до значення з ітерованої колекції, але до її індексу, тут функція enumerate стає незамінним помічником. Ця функція дозволяє вам легко отримувати доступ до індексу кожного елементу під час ітерації.
- Функція zip використовується для одночасної ітерації по кількох колекціях. Якщо вам потрібно комбінувати дані з різних джерел або виконувати операції, які залежать від декількох пов'язаних колекцій, zip дозволяє це зробити легко та елегантно.

Функція range створює послідовність чисел. Вона може бути використана різними способами:

- range(stop): Створює послідовність чисел від 0 до stop - 1.
- range(start, stop): Генерує числа від start до stop - 1.
- range(start, stop, step): Створює числа від start до stop - 1, з кроком step.

In [151]:
for i in range(5):
    print(i)

0
1
2
3
4


In [152]:
for i in range(2, 10):
    print(i)

2
3
4
5
6
7
8
9


In [153]:
for i in range(0, 10, 2):
    print(i)

0
2
4
6
8


Функція enumerate використовується для одночасного отримання індексу та значення елементів ітерованого об'єкта. Це корисно, коли вам потрібно отримати доступ до індексу елементів під час ітерації.

In [154]:
some_list = ["apple", "banana", "cherry"]
for index, value in enumerate(some_list):
    print(index, value)

0 apple
1 banana
2 cherry


Функція zip використовується для ітерації по декількох ітерованих об'єктах одночасно. Вона "застібає" елементи з кожного ітерованого об'єкта, створюючи кортежі з цих елементів.

In [155]:
list1 = ["зелене", "стигла", "червоний"]
list2 = ["яблуко", "вишня", "томат"]
for number, letter in zip(list1, list2):
    print(number, letter)

зелене яблуко
стигла вишня
червоний томат


In [157]:
list1 = [1, 2, 3]
list2 = ["a", "b", "c", "d", "e"]

for number, letter in zip(list1, list2):
    print(number, letter)

 # розмір визначається за найменшим списком   

1 a
2 b
3 c


In [158]:
numbers = {1: "one", 2: "two", 3: "three"}

In [159]:
for key in numbers:
    print(key)

1
2
3


In [160]:
for key in numbers.keys():
    print(key)

1
2
3


In [162]:
for val in numbers.values():
    print(val)

one
two
three


In [163]:
for key, value in numbers.items():
    print(key, value)

1 one
2 two
3 three


## Механізм обробки винятків

In [164]:
val = "a"
try:
    val = int(val)
except ValueError:
    print(f"val {val} is not a number")
else:
    print(val > 0)
finally:
    print("This will be printed anyway")

val a is not a number
This will be printed anyway


In [165]:
age = input("How old are you? ")
try:
    age = int(age)
    if age >= 18:
        print("You are adult.")
    else:
        print("You are infant")
except ValueError:
    print(f"{age} is not a number")

You are adult.


# Функції

In [166]:
def print_max(a, b):
    if a > b:
        print(a, "максимально")
    elif a == b:
        print(a, "дорівнює", b)
    else:
        print(b, "максимально")


print_max(3, 4)  # пряма передача значень

x = 5
y = 7
print_max(x, y)  # передача змінних у якості аргументів

4 максимально
7 максимально


## Типізація параметрів функції

В Python, починаючи з версії 3.5, ви можете використовувати "type hints" (підказки типів) для вказівки очікуваних типів параметрів. Це дозволяє коду бути більш зрозумілим та може допомогти у виявленні можливих помилок.

Попередню функцію ми можемо записати тепер наступним чином:

In [167]:
def print_max(a: int, b: int):
    if a > b:
        print(a, "максимально")
    elif a == b:
        print(a, "дорівнює", b)
    else:
        print(b, "максимально")


print_max(3, 4)  # пряма передача значень

x = 5
y = 7
print_max(x, y)  # передача змінних у якості аргументів

4 максимально
7 максимально


## Повернення результату



In [168]:
def add_numbers(num1: int, num2: int) -> int:
    sum = num1 + num2
    return sum


result = add_numbers(5, 10)
print(result)  # Виведе: 15

15


In [169]:
def greet(name: str) -> str:
    return f"Привіт, {name}!"


greeting = greet("Олексій")
print(greeting)  # Виведе: Привіт, Олексій!

Привіт, Олексій!


In [170]:
def outer_func():
    enclosing_var = "Я змінна в функції, що охоплює"

    def inner_func():
        print("Всередині вкладеної функції:", enclosing_var)

    inner_func()


outer_func()

Всередині вкладеної функції: Я змінна в функції, що охоплює


Розглянемо типову задачу, яка відображає реальну ситуацію в області торгівлі та фінансів, де потрібно часто обраховувати ціни зі знижками. Нам необхідно створити функцію для розрахунку вартості товарів з урахуванням можливої знижки.



In [171]:
def real_cost(base: int, discount: int = 0) -> int:
    return base * (1 - discount)

In [172]:
price_bread = 15
price_butter = 50
price_sugar = 60

current_price_bread = real_cost(price_bread)
current_price_butter = real_cost(price_butter, 0.05)
current_price_sugar = real_cost(price_sugar, 0.07)

In [173]:
print(f"Нова вартість хліба: {current_price_bread}")
print(f"Нова вартість масла: {current_price_butter}")
print(f"Нова вартість цукру: {current_price_sugar}")

Нова вартість хліба: 15
Нова вартість масла: 47.5
Нова вартість цукру: 55.8


У Python існує два способи реалізації змінної кількості параметрів.

- Параметр *args. Він дозволяє функції приймати довільну кількість позиційних аргументів. Аргументи, передані функції, зберігаються у вигляді кортежу.
- Параметр **kwargs. Він дозволяє функції приймати довільну кількість ключових аргументів. Але аргументи, передані функції, зберігаються вже у вигляді словника.

In [174]:
def print_all_args(*args):
    for arg in args:
        print(arg)


print_all_args(1, "hello", True)

1
hello
True


Де це може знадобиться? Насправді багато де, але ми розглянемо дуже простий та зрозумілий приклад. Нам треба написати функцію, яка буде передану будь яку кількість рядків об'єднувати в один рядок та повертати його.

In [ ]:
def concatenate(*args) -> str:
    result = ""
    for arg in args:
        result += arg
    return result


print(concatenate("Hello", " ", "world", "!"))

Параметр **kwargs використовується у визначенні функції для вказівки на те, що функція може приймати довільну кількість ключових аргументів. Назва kwargs пішла від "keyword arguments" та дозволяє передавати в функцію іменовані аргументи у вигляді словника.



In [175]:
def greet(**kwargs):
    for key, value in kwargs.items():
        print(f"{key}: {value}")


greet(name="Alice", age=25)

name: Alice
age: 25


In [176]:
def example_function(*args, **kwargs):
    print("Позиційні аргументи:", args)
    print("Ключові аргументи:", kwargs)


example_function(1, 2, 3, name="Alice", age=25)

Позиційні аргументи: (1, 2, 3)
Ключові аргументи: {'name': 'Alice', 'age': 25}


In [177]:
def greet(name, age):
    print(f"Hello {name}, you are {age} years old.")


person_info = {"name": "Alice", "age": 25}
greet(**person_info)

Hello Alice, you are 25 years old.


In [178]:
def factorial(n):
    if n == 0:  # базовий випадок
        return 1
    else:
        return n * factorial(n - 1)  # рекурсивний випадок


print(factorial(5))  # виведе 120

120


In [179]:
def fibonacci(n):
    if n <= 1:  # базовий випадок
        return n
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)  # рекурсивний випадок


print(fibonacci(10))  # виведе 55

55


# HW

In [ ]:
name = "Maryna"
age = 26

Створить програму на Python, яка розраховує вартість споживання електроенергії на основі заданого тарифу та показників лічильника.

In [94]:
rate = 1.68
value = 60
payment = rate * value
print(payment)

100.8


Створіть програму на Python, яка розраховує загальну вартість споживання електроенергії, враховуючи різні тарифи для денного та нічного часу.

In [95]:
rate = 1.68
value_day = 60
value_night = 30
payment = value_day * rate + value_night * (rate / 2)
print(payment)

126.0


Використовуючи раніше створені змінні first_name та last_name, здійсніть конкатенацію цих змінних для формування повного імені.

In [96]:
first_name = "Maryna"
last_name = "Lysenko"
full_name = first_name + " " + last_name
print(full_name)

Maryna Lysenko


In [99]:
length = 2.75
width = 1.75
area = length * width
show = (f"With width {width} and length {length} of the room, its area is equal to {area}")
print(show)

With width 1.75 and length 2.75 of the room, its area is equal to 4.8125



Булевий тип часто використовують, щоб позначити стан об'єкта. Наприклад, змінна is_active може вказувати, чи активний обліковий запис користувача, чи ні. Також акаунти користувачів часто не видаляють із бази даних, а просто встановлюють їм властивість is_delete – це називають soft delete (м'яке видалення). Використовуючи булевий тип даних у Python, встановити стан облікових записів користувачів за допомогою змінних is_active та is_delete.

In [100]:
is_active = True
is_delete = False

In [101]:
name = "Maryna"
age = 26
is_active = True
subscription = None
print(type(name))  # <class 'str'>

<class 'str'>


In [103]:
length = "2.75"
width = "1.75"
area = float(length) * float(width)
show = (
    f"With width {width} and length {length} of the room, its area is equal to {area}"
)
print(show)

With width 1.75 and length 2.75 of the room, its area is equal to 4.8125


In [104]:
name = input("Your name? ")
email = input("Your email? ")
age = int(input("Your age? "))
height = float(input("Your height? "))
is_active = bool(input("Do you want to receive messages? "))


In [105]:
length = float(input("Your length? "))
width = float(input("Your width? "))
area = length * width
show = (
    f"With width {width} and length {length} of the room, its area is equal to {area}"
)
print(show)

With width 1.5 and length 1.2 of the room, its area is equal to 1.7999999999999998


In [109]:
my_list = []
my_list.insert(0,2024)
my_list.insert(1,'Python')
my_list.insert(2,3.12)
print(my_list)

[2024, 'Python', 3.12]


In [110]:
my_list = [2024, 3.12]
some_data = ["Python"]
my_list.extend(some_data)
my_list.insert(1, "Python")
my_list.reverse()

[2024, 3.12, 'Python']
['Python', 3.12, 'Python', 2024]


In [113]:
data = dict()
data['year'] = 2024
data['lang'] = 'Python'
data['version'] = 3.12
print(data)

{'year': 2024, 'lang': 'Python', 'version': 3.12}


In [114]:
cat = {}
info = {"status": "vaccinated", "breed": True}
cat["nick"] = "Albus"
cat["age"] = 2
cat["characteristics"] = ['kind', 'playful', 'sleepy']
age = cat.get("age")
cat.update(info)
print(cat)

{'nick': 'Albus', 'age': 2, 'characteristics': ['kind', 'playful', 'sleepy'], 'status': 'vaccinated', 'breed': True}


# Q&A